kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesMockExtension.java

```diff
@@ -0,0 +1,160 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+
+/**
+ * The class that implements JUnit5 extension mechanism. You can use it directly in your JUnit test
+ * by annotating it with <code>@ExtendWith(KubernetesMockExtension.class)</code> or through
+ * <code>@EnableKubernetesMock</code> annotation
+ */
+public class KubernetesMockExtension implements AfterEachCallback, AfterAllCallback, BeforeEachCallback, BeforeAllCallback {
+
+  private KubernetesMockServer mock;
+  private DefaultKubernetesClient client;
+  private Class<? extends Client>[] extensions;
+
+  public interface SetTestClassField {
+    void apply(Object instance, Field f) throws IllegalAccessException;
+  }
+
+  @Override
+  public void afterEach(ExtensionContext context) {
+    Optional<Class<?>> optClass = context.getTestClass();
+    if (optClass.isPresent()) {
+      Class<?> testClass = optClass.get();
+      if (findField(testClass, true) == null) {
+        destroy();
+      }
+    }
+  }
+
+  @Override
+  public void afterAll(ExtensionContext context) {
+    destroy();
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) throws Exception {
+    setKubernetesClientAndMockServerFields(context, false);
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    setKubernetesClientAndMockServerFields(context, true);
+  }
+
+  protected void setFieldIfKubernetesClientOrMockServer(ExtensionContext context, boolean isStatic, Field field) throws IllegalAccessException {
+    if (extensions.length == 0) {
+      setFieldIfEqualsToProvidedType(context, isStatic, field, Client.class, (i, f) -> f.set(i, client.adapt(f.getType())));
+    } else {
+      for (Class<? extends Client> clazz : extensions) {
+        setFieldIfEqualsToProvidedType(context, isStatic, field, clazz, (i, f) -> f.set(i, client.adapt(f.getType())));
+      }
+    }
+    setFieldIfEqualsToProvidedType(context, isStatic, field, getKubernetesMockServerType(), (i, f) -> f.set(i, mock));
+  }
+
+  protected void setFieldIfEqualsToProvidedType(ExtensionContext context, boolean isStatic, Field field, Class<?> fieldType, SetTestClassField setTestClassField) throws IllegalAccessException {
+    if (fieldType.isAssignableFrom(field.getType()) && Modifier.isStatic(field.getModifiers()) == isStatic) {
+      setKubernetesClientStaticOrMemberField(context, isStatic, field, setTestClassField);
+    }
+  }
+
+  protected void initializeKubernetesClientAndMockServer(Class<?> testClass) {
+    EnableKubernetesMock a = testClass.getAnnotation(EnableKubernetesMock.class);
+    final Map<ServerRequest, Queue<ServerResponse>> responses = new HashMap<>();
+    mock = a.crud()
+      ? new KubernetesMockServer(new Context(), new MockWebServer(), responses, new KubernetesMixedDispatcher(responses), a.https())
+      : new KubernetesMockServer(a.https());
+    mock.init();
+    client = new DefaultKubernetesClient(mock.getMockConfiguration());
```

Comment:
Why is an explicit client instantiation needed?

Reply:
Because we need the functionality that is below the KubernetesClient, see the next call.  We could adapt(BaseClient.class) instead if you want to avoid what will be a deprecated call for a while.

Reply:
Then maybe we should change the return type of KubernetesMockServer#createClient to BaseClient. Then change that in the future if possible.

Reply:
Unfortunately createClient effectively became a user facing method and should not expose -client classes.

Reply:
This implies that usage of the ExternalResource won't work correctly in adapting scenarios.  We need KubernetesMockServerExtension and KubenetesServer to share the same KubernetesMockServer and client creation logic for that.